File: ports/android/app/src/main/java/dev/shivampingale/vcport/PimEstimator.kt
package dev.shivampingale.vcport

/**
 * Honest PIM helper. Not Benchmark. Not a time-to-crack estimate.
 * Matches VeraCrypt non-system volumes: HMAC KDFs use 500,000 iterations
 * at PIM 0, or PIM × 1,000. Argon2id uses a different cost; we do not
 * pretend it is the same formula.
 */
object PimEstimator {
    fun hmacIterations(pim: Int): Int = if (pim <= 0) 500_000 else pim * 1_000

    fun describe(kdf: String, pimText: String): String {
        val pim = pimText.filter { it.isDigit() }.toIntOrNull() ?: 0
        val argon = kdf.contains("Argon2", ignoreCase = true)
        return if (argon) {
            "Argon2id. PIM changes Argon2 time cost. This is not seconds-to-open and not a crack-time estimate."
        } else {
            val n = hmacIterations(pim)
            val pimBit = if (pim <= 0) "PIM 0 (VeraCrypt default)" else "PIM $pim"
            "$kdf: about ${"%,d".format(n)} header iterations ($pimBit). Not a crack-time estimate. Benchmark measures cipher speed, not this."
        }
    }
}

File: ports/ios/VCPort/PimEstimator.swift
import Foundation

/// Honest PIM helper. Not Benchmark. Not a crack-time estimate.
enum PimEstimator {
    static func hmacIterations(_ pim: Int) -> Int {
        pim <= 0 ? 500_000 : pim * 1_000
    }

    static func describe(kdf: String, pimText: String) -> String {
        let pim = Int(pimText.filter(\.isNumber)) ?? 0
        if kdf.localizedCaseInsensitiveContains("Argon2") {
            return "Argon2id. PIM changes Argon2 time cost. This is not seconds-to-open and not a crack-time estimate."
        }
        let n = hmacIterations(pim)
        let formatted = NumberFormatter.localizedString(from: NSNumber(value: n), number: .decimal)
        let pimBit = pim <= 0 ? "PIM 0 (VeraCrypt default)" : "PIM \(pim)"
        return "\(kdf): about \(formatted) header iterations (\(pimBit)). Not a crack-time estimate. Benchmark measures cipher speed, not this."
    }
}
